### PR TITLE
Improved comparator

### PR DESCRIPTION
**Pull request by divyamadala30**
Signed-off-by: Divya Madala <71040191+divyamadala30@users.noreply.github.com>

Briefly describe changes proposed in this pull request:
- Comparator now recognizes changes in libraries, qcReports and runs in SampleMetadata level

--- 
## Crossing T's and dotting I's

Please follow these checklists to help prevent any unexpected issues from being introduced by the changes in this pull request. If an item does not apply then indicate so by surrounding the line item with `~~` to strikethrough the text. See [basic writing and formatting syntax](https://docs.github.com/en/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) for more information.

### Mocked request data and data model checklist:

**Data checks:**
Updates were made to the mocked incoming request data and/or mocked published request data:
- [ ] [smile-server test data](https://github.com/mskcc/smile-server/tree/master/service/src/test/resources/data)
- [x] [smile-commons test data](https://github.com/mskcc/smile-commons/tree/master/src/test/resources/data)
- [ ] [smile-label-generator test data](https://github.com/mskcc/smile-label-generator/tree/master/src/test/resources/data)
- [ ] [smile-request-filter test data](https://github.com/mskcc/smile-request-filter/tree/master/src/test/resources/data)

**Code checks:**
- [x] The JSON comparator code been updated to handle new changes.
- [x] Unit tests were updated in relation to updates to the mocked test data.

**Describe your testing environment:**

- NATS [local, local docker, dev server, production]
- Neo4j [local, local docker, dev server, production]
- SMILE Server [local, local docker, dev server, production]
- Message publishing simulation [nats cli, docker nats cli, smile publisher tool, other (describe below)]

Other: Didn't use any of the above to test

### General checklist:
- [ ] All requested changes and comments have been resolved.
- [ ] The commit log is comprehensible. It follows [7 rules of great commit messages](http://chris.beams.io/posts/git-commit/). For most PRs a single commit should suffice, in some cases multiple topical commits can be useful. During review it is ok to see tiny commits (e.g. Fix reviewer comments), but right before the code gets merged to master or rc branch, any such commits should be squashed since they are useless to the other developers. Definitely avoid [merge commits, use rebase instead.](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)

### CODE REVIEW

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 1]
```diff
@@ -91,10 +92,54 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
+        while (itrTar.hasNext()) {
+            JsonNode sampleNode = itrTar.next();
+            if (primaryId.equals(sampleNode.get("primaryId").toString())) {
+                return sampleNode;
+            }
+        }
+        return null;
+    }
+
+    private Boolean isMatchingJsonByFieldName(JsonNode refNode, JsonNode tarNode, String fieldName) {
```
Comment (divyamadala30) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:134]: Maybe a better name for these methods...

Comment (divyamadala30) [reply src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:134]: Also should I add descriptions to these methods?

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 2]
```diff
@@ -91,10 +92,54 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
             }
+
+            JsonNode refSamplesJsonNode = mapper.readTree(referenceJson).get("samples");
+            ArrayNode refSamplesArrayNode = (ArrayNode) refSamplesJsonNode;
+            Iterator<JsonNode> itrRef = refSamplesArrayNode.elements();
+
+            // Iterating through a list of samples from referenceJson
+            while (itrRef.hasNext()) {
+                JsonNode refSampleNode = itrRef.next();
+                if (refSampleNode.has("primaryId")) {
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:103]: Maybe make this check for either `primaryId` or `igoId`?

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 3]
```diff
@@ -91,10 +92,54 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
+                    JsonNode tarSampleNode = findSampleNodeFromSampleArray(targetJson,
+                            refSampleNode.get("primaryId").toString());
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:105]: same comment here - finding a sample node by a either `primaryId` or `igoId`

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 4]
```diff
@@ -91,10 +92,54 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
+                    // Compares libraries, runs and qcReports.
+                    if (!isMatchingJsonByFieldName(refSampleNode, tarSampleNode, "qcReports")
+                            || !isMatchingJsonByFieldName(refSampleNode, tarSampleNode, "libraries")
+                            || !isMatchingJsonByFieldName(refSampleNode.get("libraries"),
+                                    tarSampleNode.get("libraries"), "runs")) {
+                        consistencyCheckStatus = Boolean.FALSE;
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:111]: I like this though - I think it makes the most sense to do it this way

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 5]
```diff
@@ -91,10 +92,54 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
 
+    private JsonNode findSampleNodeFromSampleArray(String targetJson, String primaryId)
+            throws JsonMappingException, JsonProcessingException {
+        JsonNode tarSamplesJsonNode = mapper.readTree(targetJson).get("samples");
+        ArrayNode tarSamplesArrayNode = (ArrayNode) tarSamplesJsonNode;
+        Iterator<JsonNode> itrTar = tarSamplesArrayNode.elements();
+
+        while (itrTar.hasNext()) {
+            JsonNode sampleNode = itrTar.next();
+            if (primaryId.equals(sampleNode.get("primaryId").toString())) {
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:127]: primaryId or igoId

**Review by ao508 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 6]
```diff
@@ -91,10 +94,85 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
+            while (itrRef.hasNext()) {
+                JsonNode refSampleNode = itrRef.next();
+                String primaryId = findPrimaryIdFromJsonNode(refSampleNode);
+                if (primaryId != null) {
+                    JsonNode tarSampleNode = findSampleNodeFromSampleArray(targetJson, primaryId);
+                    // Compares libraries, runs and qcReports.
+                    if (!isMatchingJsonByFieldName(refSampleNode, tarSampleNode, "qcReports")
+                            || !isMatchingJsonByFieldName(refSampleNode, tarSampleNode, "libraries")
+                            || !isMatchingJsonByFieldName(refSampleNode.get("libraries"),
+                                    tarSampleNode.get("libraries"), "runs")) {
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:112]: remove check on `runs`

File: src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java [hunk 7]
```diff
@@ -91,10 +94,85 @@ public Boolean isConsistent(String referenceJson, String targetJson, String[] ig
...
+     * @throws JsonProcessingException
+     */
+    private JsonNode convertToMapJsonNode(String fieldName, JsonNode node) throws JsonProcessingException {
+        Map<String, String> map = new HashMap<>();
+        map.put(fieldName, node.get(fieldName).toString());
+        String convertedMapAsString = mapper.writeValueAsString(map);
+        return mapper.readTree(convertedMapAsString);
+    }
+
+
```
Comment (ao508) [inline src/main/java/org/mskcc/smile/commons/impl/JsonComparatorImpl.java:170]: extra new line (also needs javadoc)

**Review by ao508 — CHANGES_REQUESTED** [review]

**Review by ao508 — APPROVED** [review]

**Pull request merged**